Create model docs for MIRGE-Com (#345)
illinois-ceesd/mirgecom: doc/discretization.rst

=======================
Discretization Strategy
=======================

.. _disc-strat:

> Literature on DG for Fluid Flow

Reviewer: I feel like this would be more readable if it also got the `\let\b=\mathbf` treatment. (Just copy from the other file.)

Author: I am really struggling with this. Copy *what* from *what* other file? I searched meshmode, grudge, mirgecom, and google for how to do this, and came up empty handed.

I tried adding `latex_elements = {"preamble": r"""\let\b=\mathbf"""}`, to no avail. It does not appear to render properly.

Author: (f1a7248) :tada:?

Reviewer 2: I vote we not let this hold up the PR unnecessarily and give the latex compiler an excuse to bug out on us.

Reviewer: #356

===============================

How to discretize the conservation equations with DG, including how to handle the required
fluxes, particularly in the viscous setting, is a current topic of research and internal
discussion.  The following references are useful:

* "The DG Book:" Nodal Discontinuous Galerkin Methods, [Hesthaven_2008]_
* The BR1 method for discretization of Navier-Stokes, [Bassi_1997]_
* NS with reactions, [Ihme_2014]_, and [Cook_2009]_
* The BR2 method, [Bassi_2000]_
* [Ayuso_2009]_

*MIRGE-Com* currently employs a strategy akin to the BR1 algorithm outlined in
[Bassi_1997]_, but with thermal terms and chemical reaction sources as outlined in
[Ihme_2014]_ and [Cook_2009]_.

Nodal DG for Navier-Stokes
==========================

The main system of equations we solve in the BR1 approach is summarized as follows:

The compressible NS equations are rewritten as the following coupled system for two
unknowns, $\mathbf{Q}$ and $\mathbf{\Sigma}$:

$$

> \mathbf{\Sigma} - \nabla{\mathbf{Q}} &= \mathbf{0}\quad \text{aux eqn}\\

Reviewer: ```suggestion
\mathbf{\Sigma} - \nabla{\mathbf{Q}} &= \mathbf{0}\tag{\text{auxiliary eqn}}\\
```

Author: Backed out and unresolved due to latex complaints.

\frac{\partial \mathbf{Q}}{\partial t} + \underbrace{\nabla\cdot\mathbf{F}^I(\mathbf{Q}) -
\nabla\cdot\mathbf{F}^V(\mathbf{Q},\mathbf{\Sigma})}_{= \nabla\cdot\mathbf{F}(\mathbf{Q},

> \mathbf{\Sigma})} &= \mathbf{S} \quad \text{primary eqn}

Reviewer: ```suggestion
\mathbf{\Sigma})} &= \mathbf{S} \tag{\text{primary eqn}}
```

Author: Backed out and unresolved due to latex CI hating these.

$$

Let $\Omega_h$ denote a collection of disjoint elements $E$. The DG method constructs
approximations $\mathbf{Q}_h$ and $\mathbf{\Sigma}_h$ to $\mathbf{Q}$ and $\mathbf{\Sigma}$,
respectively, in discontinuous finite element spaces. For any integer $k$, we define
the following discontinuous spaces of piecewise (vector-valued) polynomial functions:
$$

> \mathbf{V}^k_h &= \left\lbrace \mathbf{v} \in L^2(\Omega_h)^N \text{ such that }

Reviewer: ```suggestion
\mathbf{V}^k_h &= \left\lbrace \mathbf{v} \in L^2(\Omega_h)^N:
```

\mathbf{v}|_E \in \lbrack P^k(E) \rbrack^N, \text{ for all } E \in \Omega_h
\right\rbrace, \\

> \mathbf{W}^k_h &= \left\lbrace \mathbf{w} \in L^2(\Omega_h)^{N\times d} \text{ such that }

Reviewer: ```suggestion
\mathbf{W}^k_h &= \left\lbrace \mathbf{w} \in L^2(\Omega_h)^{N\times d}:
```

$$
where $N = d + 2 + N_s$, $d$ is the spatial dimension, and $N_s$ is the total number of
mixture species. Here, $P^k(E)$ denotes a polynomial space on $E$ consisting of functions
of degree $\leq k$. The DG formulation is obtained by multiplying by ''test functions''
$\mathbf{v}_h \in \mathbf{V}^k_h$, $\mathbf{w}_h \in \mathbf{W}^k_h$
(one for each equation repsectively) and integrating over each element.
The resulting DG problem reads as follows. Find $(\mathbf{Q}_h,
\mathbf{\Sigma}_h) \in \mathbf{V}^k_h \times \mathbf{W}^k_h$ such that, for all
$(\mathbf{v}_h, \mathbf{w}_h) \in \mathbf{V}^k_h \times \mathbf{W}^k_h$, we have:

$$
\sum_{E\in\Omega_h} \left\lbrack \int_E \mathbf{v}_h\cdot\frac{\partial \mathbf{Q}_h}
{\partial t} d\Omega + \oint_{\partial E} \mathbf{v}_h\mathbf{h} d\sigma - \int_E \nabla
\mathbf{v}_h\cdot\mathbf{F}(\mathbf{Q}_h, \mathbf{\Sigma}_h)d\Omega\right\rbrack &=
\sum_{E\in\Omega_h} \int_E \mathbf{v}_h\cdot\mathbf{S}_h d\Omega, \\
\sum_{E\in\Omega_h}\left\lbrack \int_E \mathbf{w}_h\cdot \mathbf{\Sigma}_h d\Omega -
\oint_{\partial E} \mathbf{w}_h\cdot\mathbf{H}_s d\sigma + \int_E \nabla\mathbf{w}_h\cdot
\mathbf{Q}_h d\Omega\right\rbrack &= 0,
$$

> where $\mathbf{h} = \mathbf{h}_e - \mathbf{h}_v$ is a *numerical flux* consisting of a

Reviewer: I'm used to reading numerical fluxes marked up with an asterisk and including the normal, like `$(h \cdot n)^*$`. How do you feel about that?

Author: I like that just fine.  The thing is, the BR1 reference uses plain-ol' H,h for numerical fluxes and they reserve h^*, H^* for the domain boundary-specific versions of those functions.

Reviewer: Fair enough. Let's stick with the BR1 notation for now.

> *invicid* and *viscous* component respectively, and $\mathbf{H}_s$ is the

Reviewer: ```suggestion
*inviscid* and *viscous* component respectively, and $\mathbf{H}_s$ is the
```

> *gradient numerical flux* for the auxiliary equation. Here, we use the subscript "h" to

Reviewer: ```suggestion
*gradient numerical flux* for the auxiliary equation. Here, we use the subscript "$h$" to
```

> denote discretized quantities.

Reviewer: ```suggestion
denote discretized quantities. See below for more on these functions.
```

Since $\mathbf{F}^I(\mathbf{Q}_h)\cdot\mathbf{n}$ is discontinuous, the quantities are
allowed to vary on either side of a shared element boundary. That is, $\mathbf{F}^I
(\mathbf{Q}^+_h)\cdot\mathbf{n}^+ \neq \mathbf{F}^I(\mathbf{Q}^-_h)\cdot\mathbf{n}^-$,

> where $\mathbf{n}^\pm$ denotes the ''cell-local'' (or interior) normal with respect to the

Reviewer: ```suggestion
where $\mathbf{n}^+$ denotes the exterior normal with respect to the
```
? (or explain both signs)
cc @thomasgibson

Reviewer 2: I clarified in https://github.com/illinois-ceesd/mirgecom/pull/345/commits/8554dde3589abe8e3f97e437f0c178764e7b30a9

> element $E^\pm$ which share a face: $\partial E^+ \cap \partial E^- \neq \emptyset$.

Reviewer: ```suggestion
elements $E^\pm$ which share a face: $\partial E^+ \cap \partial E^- \neq \emptyset$.
```

Similarly for $\mathbf{F}^V(\mathbf{Q}_h, \mathbf{\Sigma}_h)\cdot\mathbf{n}$ and
$\mathbf{Q}_h\mathbf{n}$.

Expanding out the trial and test functions (component-wise) in terms of the local
element basis in each element:
$$
\mathbf{Q}(\mathbf{x}, t)_h|_E = \sum_{i=1}^n \mathbf{Q}_i(t)\phi_i^k(\mathbf{x}), \quad
\mathbf{\Sigma}(\mathbf{x})_h|_E = \sum_{i=1}^n \mathbf{\Sigma}_i\phi_i^k(\mathbf{x}), \\
\mathbf{v}(\mathbf{x})_h|_E = \sum_{i=1}^n \mathbf{v}_i\phi_i^k(\mathbf{x}), \quad
\mathbf{w}(\mathbf{x})_h|_E = \sum_{i=1}^n \mathbf{w}_i\phi_i^k(\mathbf{x}),
$$

> allows us to obtain a set of algebraic equations for the prognostic state $\mathbf{Q}_h$ and

Reviewer: Is "prognostic" a word that's used outside of weather forecasting? Never heard it before @thomasgibson showed up. :)

Author: Yes. It's a word you never want to hear your M.D. say.

Reviewer 2: I removed the medical terminology in https://github.com/illinois-ceesd/mirgecom/pull/345/commits/8554dde3589abe8e3f97e437f0c178764e7b30a9 :)

the auxiliary gradient variable $\mathbf{\Sigma}_h$. That is, for each

> $j = 1, \cdots, \dim P^k$, we have:

Reviewer: ```suggestion
$j = 1, \dots, \dim P^k$, we have:
```

    {2}\left(\mathbf{Q}_h^+ - \mathbf{Q}_h^-\right)\mathbf{n},
    $$
    where $\lambda$ is the characteristic max wave-speed of the fluid. Numerical fluxes
    which penalize the ''jump'' of the state $\left(\mathbf{Q}_h^+ - \mathbf{Q}_h^-\right)
    \mathbf{n}$ act as an additional source of dissipation, which has a stabilizing effect
    on the numerics.

* Take $\mathbf{h}_v(\mathbf{Q}_h^+, \mathbf{\Sigma}_h^+, \mathbf{Q}_h^-,
  \mathbf{\Sigma}_h^-; \mathbf{n})$ to be defined in a ''centered'' (average) way:
  $$
  \mathbf{h}_v(\mathbf{Q}_h^+, \mathbf{\Sigma}_h^+, \mathbf{Q}_h^-, \mathbf{\Sigma}_h^-;
  \mathbf{n}) = \frac{1}{2}\left(\mathbf{F}^V(\mathbf{Q}_h^+, \mathbf{\Sigma}_h^+) +
  \mathbf{F}^V(\mathbf{Q}_h^-, \mathbf{\Sigma}_h^-)\right)\cdot\mathbf{n}
  $$

* And similarly for the gradient flux:
  $$
  \mathbf{H}_s(\mathbf{Q}_h^+, \mathbf{Q}_h^- ; \mathbf{n}) = \frac{1}{2}\left(
  \mathbf{Q}_h^+ + \mathbf{Q}_h^-\right)\mathbf{n}
  $$

Domain boundary considerations
------------------------------

What happens when $\partial E \cap \partial\Omega \neq \emptyset$?

In DG, numerical fluxes are not only responsible for handling the flow of information

> between adjacent cells, but they also enforce information at the boundaries.

Reviewer: ```suggestion
between adjacent cells, but they also enforce information flow at the boundaries.
```

We denote the *boundary fluxes* as $\mathbf{h}^*_e(\mathbf{Q}_{bc})$,
$\mathbf{h}^*_v(\mathbf{Q}_{bc}$, $\mathbf{\Sigma}_{bc})$, and
$\mathbf{H}^*_s(\mathbf{Q}_{bc})$, where $\mathbf{Q}_{bc}$, $\mathbf{\Sigma}_{bc}$ denote
boundary conditions imposed on the state, and the gradient of the state respectively.

For all $\partial E \cap \partial\Omega$ there is no $+$ side to consider; just the
interior state ($-$ side) and the prescribed boundary conditions $\mathbf{Q}_{bc},
\mathbf{\Sigma}_{bc}$.

> * At solid walls:

Reviewer: ```suggestion
Solid walls
^^^^^^^^^^^
```
(I prefer that over the bullets, also below)

Author: (fa4f487)

  $\mathbf{F}^I(\mathbf{Q}_{bc})\cdot\mathbf{n}$ (since $\mathbf{V}\cdot\mathbf{n} = 0$).

  * The viscous boundary flux is computed as:
    $$
    \mathbf{h}^*_v(\mathbf{Q}_{bc}, \mathbf{\Sigma}_{bc}) = \mathbf{F}_V(\mathbf{Q}_{bc},
    \mathbf{\Sigma}_{bc})\cdot\mathbf{n},
    $$
    where $\mathbf{Q}_{bc}$ are the same values used to prescribe $\mathbf{h}^*_e$.

  * If there are no conditions on $\nabla\mathbf{Q}\cdot\mathbf{n}$, then
    $$
    \mathbf{\Sigma}_{bc} = \mathbf{\Sigma}_h^-.
    $$

  Otherwise, $\mathbf{\Sigma}_{bc}$ will need to be modified accordingly.

* At inflow/outflow boundaries:

  $$
  \mathbf{h}^*_e(\mathbf{Q}_{bc}) = \mathbf{h}_e(\mathbf{Q}_{bc}, \mathbf{Q}^-_{h};
  \mathbf{n}).
  $$

  * $\mathbf{Q}_{bc}$ is also used to define the gradient boundary flux:
    $$
    \mathbf{H}^*_s(\mathbf{Q}_{bc}) = \mathbf{Q}_{bc}\mathbf{n}.
    $$

  * The viscous boundary flux is evaluated as:
    $$
    \mathbf{h}^*_v = \mathbf{h}_v(\mathbf{Q}_{bc}, \mathbf{\Sigma}_h^-, \mathbf{Q}_h^-,
    \mathbf{\Sigma}_h^-; \mathbf{n}),
    $$
    where $\mathbf{Q}_{bc}$ are the same values used for $\mathbf{h}^*_e$.

> Second order terms on the viscous RHS

Reviewer: ```suggestion
Second-order terms on the viscous RHS
```

The resulting DG problem reads as follows. Find $(\b{Q}_h,
\b{\Sigma}_h) \in \b{V}^k_h \times \b{W}^k_h$ such that, for all
$(\b{v}_h, \b{w}_h) \in \b{V}^k_h \times \b{W}^k_h$, we have:

$$
\sum_{E\in\Omega_h} \left\lbrack \int_E \b{v}_h\cdot\frac{\partial \b{Q}_h}
{\partial t} d\Omega + \oint_{\partial E} \b{v}_h\b{h} d\sigma - \int_E \nabla
\b{v}_h\cdot\b{F}(\b{Q}_h, \b{\Sigma}_h)d\Omega\right\rbrack &=
\sum_{E\in\Omega_h} \int_E \b{v}_h\cdot\b{S}_h d\Omega, \\
\sum_{E\in\Omega_h}\left\lbrack \int_E \b{w}_h\cdot \b{\Sigma}_h d\Omega -
\oint_{\partial E} \b{w}_h\cdot\b{H}_s d\sigma + \int_E \nabla\b{w}_h\cdot
\b{Q}_h d\Omega\right\rbrack &= 0,
$$

where $\b{h} = \b{h}_e - \b{h}_v$ is a *numerical flux* consisting of a
*inviscid* and *viscous* component respectively, and $\b{H}_s$ is the
*gradient numerical flux* for the auxiliary equation. Here, we use the subscript "$h$" to
denote discretized quantities. See below for more on these functions.

Since $\b{F}^I(\b{Q}_h)\cdot\b{n}$ is discontinuous, the quantities are
allowed to vary on either side of a shared element boundary. That is, $\b{F}^I
(\b{Q}^+_h)\cdot\b{n}^+ \neq \b{F}^I(\b{Q}^-_h)\cdot\b{n}^-$.

> Here, $\b{n}^+$ and $\b{n}^-$ denote the ''cell-local'' (outward pointing) normals

Reviewer: ```suggestion
Here, $\b{n}^+$ and $\b{n}^-$ denote the ''cell-local'' (outward pointing) unit normals
```

\left\lbrack\int_E \nabla\phi_j^k\cdot\left(\b{F}^V(\b{Q}_h, \b{\Sigma}_h) -
\b{F}^I(\b{Q}_h)\right)d\Omega\right\rbrack \\
&- \sum_{E\in\Omega_h}\left\lbrack\oint_{\partial E}\phi_j^k \b{h}_e(\b{Q}_h^+,
\b{Q}^-_h; \b{n}) d\sigma + \oint_{\partial E} \phi_j^k \b{h}_v(\b{Q}_h^+,
\b{\Sigma}_h^+, \b{Q}_h^-, \b{\Sigma}_h^-; \b{n}) d\sigma\right\rbrack \\
&+ \sum_{E\in\Omega_h} \int_E \phi_j^k\b{S}_h d\Omega, \\
\sum_{E\in\Omega_h}\int_E\phi_j^k \b{\Sigma}_h d\Omega &= \sum_{E\in\Omega_h}\left\lbrack
\oint_{\partial{E}}\phi_j^k \b{H}_s(\b{Q}^+_h, \b{Q}_h^-; \b{n}) d\sigma -
\int_E\nabla\phi^k_j\cdot\b{Q}_h d\Omega\right\rbrack.
$$

Numerical fluxes
================

To account for the discontinuities at element faces, DG employs numerical fluxes, which are
enforced to be singled-valued, but are functions of both $\pm$ states:

$$
\b{h}_e(\b{Q}_h^+, \b{Q}^-_h; \b{n}) &\approx \b{F}^I(\b{Q}_h)
\cdot\b{n}, \\
\b{h}_v(\b{Q}_h^+, \b{\Sigma}_h^+, \b{Q}_h^-, \b{\Sigma}_h^-;
\b{n}) &\approx \b{F}^V(\b{Q}_h, \b{\Sigma}_h)\cdot\b{n}\\
\b{H}_s(\b{Q}^+_h, \b{Q}_h^-; \b{n}) &\approx \b{Q}_h\b{n}.
$$

Choices of numerical fluxes corresponding to BR1
------------------------------------------------

> Numerical inviscid flux

Reviewer: ```suggestion
Inviscid numerical flux
```

^^^^^^^^^^^^^^^^^^^^^^^

Typical choices for $\b{h}_e(\b{Q}_h^+, \b{Q}^-_h; \b{n})$ include,
but are not limited to:

* Local Lax-Friedrichs (LLF)
* Roe
* Engquist-Osher
* HLLC

|mirgecom| currently uses LLF, which is implemented as follows:
$$
\b{h}_{e}(\b{Q}_h^+, \b{Q}^-_h; \b{n}) = \frac{1}{2}\left(
\b{F}^{I}(\b{Q}_h^+)+\b{F}^{I}(\b{Q}_h^-)\right) - \frac{\lambda}
{2}\left(\b{Q}_h^+ - \b{Q}_h^-\right)\b{n},
$$
where $\lambda$ is the characteristic max wave-speed of the fluid. Numerical fluxes
which penalize the ''jump'' of the state $\left(\b{Q}_h^+ - \b{Q}_h^-\right)
\b{n}$ act as an additional source of dissipation, which has a stabilizing effect
on the numerics.

> Numerical viscous flux

Reviewer: ```suggestion
Viscous numerical flux
```

^^^^^^^^^^^^^^^^^^^^^^
Take $\b{h}_v(\b{Q}_h^+, \b{\Sigma}_h^+, \b{Q}_h^-, \b{\Sigma}_h^-; \b{n})$ to be
defined in a ''centered'' (average) way:
$$
\b{h}_v(\b{Q}_h^+, \b{\Sigma}_h^+, \b{Q}_h^-, \b{\Sigma}_h^-;
\b{n}) = \frac{1}{2}\left(\b{F}^V(\b{Q}_h^+, \b{\Sigma}_h^+) +
\b{F}^V(\b{Q}_h^-, \b{\Sigma}_h^-)\right)\cdot\b{n}
$$

> Numerical gradient flux

Reviewer: ```suggestion
Gradient numerical flux
```

^^^^^^^^^^^^^^^^^^^^^^^
And similarly for the gradient flux:
$$
\b{H}_s(\b{Q}_h^+, \b{Q}_h^- ; \b{n}) = \frac{1}{2}\left(
\b{Q}_h^+ + \b{Q}_h^-\right)\b{n}
$$

> Domain boundary considerations

Reviewer: `\b n` is used in the BC section but not defined.

Author: Sharpened a bit in (f433eba).

------------------------------

What happens when $\partial E \cap \partial\Omega \neq \emptyset$?

In DG, numerical fluxes are not only responsible for handling the flow of information
between adjacent cells, but they also enforce information flow at the boundaries.

We denote the *boundary fluxes* as $\b{h}^*_e(\b{Q}_{bc})$,
$\b{h}^*_v(\b{Q}_{bc}$, $\b{\Sigma}_{bc})$, and
$\b{H}^*_s(\b{Q}_{bc})$, where $\b{Q}_{bc}$, $\b{\Sigma}_{bc}$ denote

> boundary conditions imposed on the state, and the gradient of the state respectively.

Reviewer: Where doe these come from? What are they for solid walls?

Author: Is it sufficient to mention particular boundary treatments are described in the sections down below (0cc222c)?

For all $\partial E \cap \partial\Omega$ there is no $+$ side to consider; just the
interior state ($-$ side) and the prescribed boundary conditions $\b{Q}_{bc},

> \b{\Sigma}_{bc}$.

Reviewer: I don't like this phrasing very much. I still think that weak imposition (via prescribing +-side state) is the better way to do BCs in DG. I'm OK with forgoing that for practicality's sake, but the phrasing here suggests that the approach doesn't exist... which rubs me the wrong way.

Author: How about (0cc222c).

Solid walls
^^^^^^^^^^^

Inviscid boundary flux
""""""""""""""""""""""
$\b{h}^*_e$ is equal to the (interior; - side) pressure contribution of
$\b{F}^I(\b{Q}_{bc})\cdot\b{n}$
(since $\b{V}\cdot\b{n} = 0$).

Viscous boundary flux
"""""""""""""""""""""
$$
\b{h}^*_v(\b{Q}_{bc}, \b{\Sigma}_{bc}) = \b{F}_V(\b{Q}_{bc},
\b{\Sigma}_{bc})\cdot\b{n},
$$
where $\b{Q}_{bc}$ are the same values used to prescribe $\b{h}^*_e$.

Gradient boundary flux
""""""""""""""""""""""
If there are no conditions on $\nabla\b{Q}\cdot\b{n}$, then:
$$
\b{\Sigma}_{bc} = \b{\Sigma}_h^-.
$$
Otherwise, $\b{\Sigma}_{bc}$ will need to be modified accordingly.

Inflow/outflow boundaries
^^^^^^^^^^^^^^^^^^^^^^^^^
Inviscid boundary flux
""""""""""""""""""""""
$$
\b{h}^*_e(\b{Q}_{bc}) = \b{h}_e(\b{Q}_{bc}, \b{Q}^-_{h};
\b{n}).
$$

Viscous boundary flux
"""""""""""""""""""""
$$
\b{h}^*_v = \b{h}_v(\b{Q}_{bc}, \b{\Sigma}_h^-, \b{Q}_h^-,
\b{\Sigma}_h^-; \b{n}),
$$
where $\b{Q}_{bc}$ are the same values used for $\b{h}^*_e$.

Gradient boundary flux
""""""""""""""""""""""
$\b{Q}_{bc}$ is also used to define the gradient boundary flux:
$$
\b{H}^*_s(\b{Q}_{bc}) = \b{Q}_{bc}\b{n}.
$$

> Second-order terms on the viscous RHS

Reviewer: I'm not totally sure why this is here... it feels more like internal notes. It's also not discrete, which makes this seemingly a bit out of place in the discretization section?

Author: Let's remove it.

Author: Had trouble letting go of it I guess. (5e939db), (f826ea1)

=====================================

> This section breaks out explicit component versions of the 2nd order terms on the RHS to

Reviewer: ```suggestion
This section breaks out explicit component versions of the second order terms on the RHS to
```

help guide a discussion about alternate approaches to discretization in which a generic
diffusion operator could potentially be responsible for integrating some of these terms.

The viscous fluxes $\b{F}^{V}$ are proportional to gradients of the fluid state

> variables, introducing 2nd order terms on the RHS of the conservation equations. These 2nd

Reviewer: ```suggestion
variables, introducing second order terms on the RHS of the conservation equations. These second
```
